File: sdk/typescript-guest/capabilities/streaming.ts
import * as raw from 'latent:http/streaming@0.3.0';
import { Owner, drop } from './owner.js';
import { call } from './result.js';
export type { Request, Header, Method, HttpError } from 'latent:http/streaming@0.3.0';
export type Upload = Owner<raw.Upload, 'http-upload'>;
export type Body = Owner<raw.Body, 'http-body'>;
export type Chunk = Owner<raw.Chunk, 'http-chunk'>;
export type Response = Omit<raw.Response, 'body'> & { body: Body };
const errors: readonly raw.HttpError['tag'][] = ['invalid-url', 'invalid-request', 'permission-denied',
  'request-too-large', 'response-too-large', 'deadline-exceeded', 'cancelled', 'budget-exhausted',
  'dns-failed', 'tls-failed', 'connection-failed', 'unavailable', 'uncertain', 'invalid-state',
  'unexpected-eof', 'unsupported-encoding'];
function invoke<T>(operation: () => T) { return call<T, raw.HttpError>(operation, errors); }
export function open(request: raw.Request) {
  return invoke((): Upload => new Owner(raw.open(request), drop));
}
export function write(upload: Upload, bytes: Uint8Array) {
  return invoke(() => upload.borrow(value => raw.write(value, bytes)));
}
export function finish(upload: Upload) {
  return invoke((): Response => {
    const response = upload.consume(value => raw.finish(value));
    return { ...response, body: new Owner(response.body, drop) };
  });
}
export function read(body: Body, maximumBytes: number) {
  return invoke((): Chunk | undefined => body.borrow(value => {
    const chunk = raw.read(value, maximumBytes);
    return chunk === undefined ? undefined : new Owner(chunk, drop);
  }));
}
/** A chunk remains charged until close, including after materialization. */
export function chunkBytes(chunk: Chunk) { return invoke(() => chunk.borrow(raw.chunkBytes)); }
export function trailers(body: Body) { return invoke(() => body.borrow(raw.trailers)); }
export function abortUpload(upload: Upload) { return invoke(() => upload.consume(raw.abortUpload)); }
export function abortBody(body: Body) { return invoke(() => body.consume(raw.abortBody)); }

File: sdk/rust-guest/src/streaming.rs
//! Owned HTTP resources with explicit asynchronous operations. Dropping an upload
//! or body aborts its I/O through the generated resource destructor; held chunks
//! keep their own charge. No operation retries or renews the activation deadline.
use crate::bindings::streaming as raw;
pub use raw::{Header, HttpError, Method, Request};

#[must_use = "finish or abort the upload; dropping it aborts I/O"]
pub struct Upload(raw::Upload);

impl Upload {
    pub async fn open(request: Request) -> Result<Self, HttpError> {
        raw::open(request).await.map(Self)
    }

    pub async fn write(&mut self, bytes: Vec<u8>) -> Result<(), HttpError> {
        raw::write(&self.0, bytes).await
    }

    pub async fn finish(self) -> Result<Response, HttpError> {
        let response = raw::finish(self.0).await?;
        Ok(Response {
            status: response.status,
            headers: response.headers,
            body_media_type: response.body_media_type,
            body: Body(response.body),
        })
    }

    pub async fn abort(self) -> Result<(), HttpError> {
        raw::abort_upload(self.0).await
    }
}

pub struct Response {
    pub status: u16,
    pub headers: Vec<Header>,
    pub body_media_type: Option<String>,
    pub body: Body,
}

#[must_use = "read or abort the body; dropping it aborts I/O"]
pub struct Body(raw::Body);

impl Body {
    /// `None` means verified EOF; an incomplete response remains a typed error.
    pub async fn read(&mut self, maximum_bytes: u32) -> Result<Option<Chunk>, HttpError> {
        raw::read(&self.0, maximum_bytes)
            .await
            .map(|chunk| chunk.map(Chunk))
    }

    /// Read once, after verified EOF. Premature or repeated calls retain the
    /// host's `InvalidState` error.
    pub async fn trailers(&mut self) -> Result<Vec<Header>, HttpError> {
        raw::trailers(&self.0).await
    }

    pub async fn abort(self) -> Result<(), HttpError> {
        raw::abort_body(self.0).await
    }
}

/// A host resource, released by generated Component Model `Drop`.
pub struct Chunk(raw::Chunk);

impl Chunk {
    pub async fn bytes(self) -> Result<Vec<u8>, HttpError> {
        raw::chunk_bytes(&self.0).await
    }
}
